crates/hir/src/source_analyzer.rs

```diff
@@ -459,24 +459,68 @@ pub(crate) fn resolve_hir_path(
     resolve_hir_path_(db, resolver, path, false)
 }
 
+fn resolve_assoc_type(
+    db: &dyn HirDatabase,
+    ty: TypeNs,
+    unresolved: PathSegment,
+) -> Option<PathResolution> {
+    let traits_from_env = match ty {
+        TypeNs::GenericParam(param_id) => {
+            let predicates = db.generic_predicates_for_param(param_id);
+            let mut traits: Vec<_> = predicates
```

Comment:
Can we avoid the allocation here?

Reply:
There might be, the question is how to unify the types of the two branches that return traits, I'm not sure how we can change that(that is in `associated_type_shorthand_candidates` which looks basically the same as this)